File: src/main/java/com/laamella/code_state_machine/util/DslStateMachineBuilder.java
package com.laamella.code_state_machine.util;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.laamella.code_state_machine.Action;
import com.laamella.code_state_machine.Precondition;
import com.laamella.code_state_machine.StateMachine;

/**
 * A pretty "DSL" builder for a state machine.
 */
public class DslStateMachineBuilder<T, E, P extends Comparable<P>> {
	private static final Logger log = LoggerFactory.getLogger(StateMachine.class);

	public class DefiningState {
		private Set<T> sourceStates = new HashSet<T>();

		public DefiningState(Set<T> sourceStates) {
			this.sourceStates = sourceStates;
		}

		public DefiningState except(final T... states) {
			for (final T state : states) {
				sourceStates.remove(state);
			}
			return this;
		}

		public DefiningState onExit(final Action action) {
			for (final T sourceState : sourceStates) {
				builder.setExitAction(sourceState, action);
			}
			return this;
		}

		public DefiningState onEntry(final Action action) {
			for (final T sourceState : sourceStates) {
				log.debug("Create entry action for {} ({})", sourceState, action);
				builder.setEntryAction(sourceState, action);
			}
			return this;
		}

		public DefiningState isAnEndState() {
			for (final T state : sourceStates) {
				builder.addEndState(state);
			}
			return this;
		}

		public DefiningState isAStartState() {
			for (final T state : sourceStates) {
				builder.addStartState(state);
			}
			return this;
		}

		public DefiningState areEndStates() {
			return isAnEndState();
		}

		public DefiningState areStartStates() {
			return isAStartState();
		}

		public DefiningTransition when(final Precondition<E> precondition) {
			return new DefiningTransition(sourceStates, precondition);
		}

		public DefiningTransition when(final E... events) {
			return new DefiningTransition(sourceStates, is(events));
		}

	}

	public class DefiningTransition {
		private Precondition<E> storedPrecondition;
		private Action action = nothing();
		private final Set<T> sourceStates;
		private P priority = defaultPriority;

		public DefiningTransition(Set<T> sourceStates, Precondition<E> precondition) {
			this.sourceStates = sourceStates;
			this.storedPrecondition = precondition;
		}

		public DefiningTransition action(final Action action) {
			this.action = action;
			return this;
		}

		public DefiningState then(final T destinationState) {
			return transition(destinationState, storedPrecondition, action, priority);
		}

		public DefiningState transition(final T destinationState, final Precondition<E> precondition,
				final Action action, P priority) {
			for (final T sourceState : sourceStates) {
				builder.addTransition(new BasicTransition<T, E, P>(sourceState, destinationState, precondition, action,
						priority));
			}
			return new DefiningState(sourceStates);
		}

		public DefiningTransition withPrio(P priority) {
			this.priority = priority;
			return this;
		}
	}

	private final StateMachine.Builder<T, E, P> builder = new StateMachine.Builder<T, E, P>();
	private final P defaultPriority;

	public DslStateMachineBuilder(P defaultPriority) {
		this.defaultPriority = defaultPriority;
	}

	public StateMachine<T, E, P> buildMachine() {
		return builder.build();
	}

	@SuppressWarnings("unchecked")
	public DefiningState state(final T state) {
		return states(state);
	}

	public DefiningState states(final T... states) {
		return new DefiningState(new HashSet<T>(Arrays.asList(states)));
	}

	public static <E> Precondition<E> always() {
		return new AlwaysPrecondition<E>();
	}

	public static <E> Precondition<E> never() {
		return new NeverPrecondition<E>();
	}

	public static <E> Precondition<E> after(final long milliseconds) {
		return new AfterPrecondition<E>(milliseconds);
	}

	public static <E> Precondition<E> is(final E... events) {
		assert events != null;
		assert events.length != 0;

		if (events.length == 1) {
			final E singleEvent = events[0];
			return new SingleEventMatchPrecondition<E>(singleEvent);
		}

		return new MultiEventMatchPrecondition<E>(events);
	}

	public static Action nothing() {
		return new NoAction();
	}

}

File: kotlin/src/main/kotlin/com/laamella/kode_state_machine/builder/DslStateMachineBuilder.kt
package com.laamella.kode_state_machine.builder

import com.laamella.kode_state_machine.Action
import com.laamella.kode_state_machine.Actions
import com.laamella.kode_state_machine.Condition
import com.laamella.kode_state_machine.Conditions
import com.laamella.kode_state_machine.StateMachine
import com.laamella.kode_state_machine.Transition
import com.laamella.kode_state_machine.action.LogAction
import com.laamella.kode_state_machine.condition.AfterCondition
import com.laamella.kode_state_machine.condition.AlwaysCondition
import com.laamella.kode_state_machine.condition.MultiEventMatchCondition
import com.laamella.kode_state_machine.condition.NeverCondition
import com.laamella.kode_state_machine.condition.SingleEventMatchCondition
import com.laamella.kode_state_machine.condition.StatesActiveCondition
import com.laamella.kode_state_machine.condition.StatesInactiveCondition
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.*

/**
 * A pretty "DSL" builder for a state machine.
 */
abstract class DslStateMachineBuilder<T, E, P : Comparable<P>>(private val defaultPriority: P) :
    StateMachineBuilder<T, E, P> {
    inner class DefiningState(
        private val sourceStates: MutableSet<T>,
        private val internals: StateMachine<T, E, P>.Internals
    ) {

        fun except(vararg states: T): DefiningState {
            for (state in states) {
                sourceStates.remove(state)
            }
            return this
        }

        fun onExit(vararg action: Action): DefiningState {
            for (sourceState in sourceStates) {
                internals.addExitActions(sourceState, *action)
            }
            return this
        }

        fun onEntry(vararg action: Action): DefiningState {
            for (sourceState in sourceStates) {
                log.debug("Create entry action for {} ({})", sourceState, action)
                internals.addEntryActions(sourceState, *action)
            }
            return this
        }

        val isAnEndState: DefiningState
            get() {
                for (state in sourceStates) {
                    internals.addEndState(state)
                }
                return this
            }

        val isAStartState: DefiningState
            get() {
                for (state in sourceStates) {
                    internals.addStartState(state)
                }
                return this
            }

        fun areEndStates(): DefiningState {
            return this.isAnEndState
        }

        fun areStartStates(): DefiningState {
            return this.isAStartState
        }

        fun `when`(vararg condition: Condition<E>): DefiningTransition {
            return DefiningTransition(sourceStates, Conditions(*condition), internals)
        }

        // This method exists only to suppress warnings about varargs.
        fun `when`(condition: Condition<E>): DefiningTransition {
            return DefiningTransition(sourceStates, Conditions(condition), internals)
        }

        fun `when`(vararg events: E): DefiningTransition {
            return DefiningTransition(sourceStates, `is`(*events), internals)
        }
    }

    inner class DefiningTransition(
        private val sourceStates: MutableSet<T>,
        private val conditions: Conditions<E>,
        private val internals: StateMachine<T, E, P>.Internals
    ) {
        private val actions = Actions()
        private var priority: P = defaultPriority

        fun action(action: Action): DefiningTransition {
            this.actions.add(action)
            return this
        }

        fun then(destinationState: T): DefiningState {
            return transition(destinationState, conditions, priority, actions)
        }

        fun transition(
            destinationState: T,
            storedConditions2: Conditions<E>,
            priority: P,
            actions: Actions
        ): DefiningState {
            this.actions.add(actions)
            for (sourceState in sourceStates) {
                internals.addTransition(
                    Transition(
                        sourceState,
                        destinationState,
                        storedConditions2,
                        priority,
                        this.actions
                    )
                )
            }
            return DefiningState(sourceStates, internals)
        }

        fun transition(
            destinationState: T,
            condition: Condition<E>,
            priority: P,
            vararg actions: Action
        ): DefiningState {
            return transition(destinationState, Conditions<E>(condition), priority, Actions(*actions))
        }

        fun withPrio(priority: P): DefiningTransition {
            this.priority = priority
            return this
        }
    }

    private var machine: StateMachine<T, E, P>? = null

    override fun build(newMachine: StateMachine<T, E, P>): StateMachine<T, E, P> {
        machine = newMachine
        executeBuildInstructions()
        return machine!!
    }

    protected abstract fun executeBuildInstructions()

    override fun build(): StateMachine<T, E, P> {
        return build(StateMachine())
    }

    fun state(state: T): DefiningState {
        return states(state)
    }

    fun states(vararg states: T): DefiningState {
        return DefiningState(HashSet(Arrays.asList(*states)), machine!!.Internals())
    }

    fun active(vararg statesThatMustBeActive: T): Condition<E> {
        return StatesActiveCondition(machine!!, *statesThatMustBeActive)
    }

    fun inactive(vararg statesThatMustBeInactive: T): Condition<E> {
        return StatesInactiveCondition(machine!!, *statesThatMustBeInactive)
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(DslStateMachineBuilder::class.java)

        fun <E> always(): Condition<E> {
            return AlwaysCondition()
        }

        fun <E> never(): Condition<E> {
            return NeverCondition()
        }

        fun <E> after(milliseconds: Long): Condition<E> {
            return AfterCondition(milliseconds)
        }

        fun <E> `is`(vararg events: E): Conditions<E> {
            assert(events.isNotEmpty())

            if (events.size == 1) {
                val singleEvent: E = events[0]
                return Conditions(SingleEventMatchCondition<E>(singleEvent))
            }

            return Conditions(MultiEventMatchCondition<E>(*events))
        }

        fun log(logText: String): Action {
            return LogAction(logText)
        }
    }
}
